dipy/reconst/qtdmri.py

```diff
@@ -0,0 +1,1952 @@
+                    (ux / (uy * uz)) *
+                    T_mat[x[i], x[k]] * U_mat[y[i], y[k]] * U_mat[z[i], z[k]]
+                )
+            LR[i, k] = LR[k, i] = val
+    return LR
+
+
+def part23_iso_reg_matrix_q(ind_mat, us):
+    """Partial spherical spatial Laplacian regularization matrix following the
+    equation below Eq. (C4) in [1].
+
+    References
+    ----------
+    .. [1] Fick, Rutger HJ, et al. "Non-Parametric GraphNet-Regularized
+        Representation of dMRI in Space and Time", Medical Image Analysis,
+        2017.
+    """
+    n_elem = ind_mat.shape[0]
+
+    LR = np.zeros((n_elem, n_elem))
+
+    for i in range(n_elem):
+        for k in range(i, n_elem):
+            if ind_mat[i, 1] == ind_mat[k, 1] and \
+               ind_mat[i, 2] == ind_mat[k, 2]:
+                ji = ind_mat[i, 0]
+                jk = ind_mat[k, 0]
+                l = ind_mat[i, 1]
+                if ji == (jk + 1):
+                    LR[i, k] = LR[k, i] = (
+                        2 ** (-l) * -gamma(3 / 2.0 + jk + l) / gamma(jk)
+                    )
+                elif ji == jk:
+                    LR[i, k] = LR[k, i] = 2 ** (-(l+1)) *\
+                        (1 - 4 * ji - 2 * l) *\
+                        gamma(1 / 2.0 + ji + l) / gamma(ji)
+                elif ji == (jk - 1):
+                    LR[i, k] = LR[k, i] = 2 ** (-l) *\
+                        -gamma(3 / 2.0 + ji + l) / gamma(ji)
+    return LR / us
+
+
+def part4_reg_matrix_q(ind_mat, U_mat, us):
+    """Partial cartesian spatial Laplacian regularization matrix following
+    equation Eq. (B2) in [1].
+
+    References
+    ----------
+    .. [1] Fick, Rutger HJ, et al. "Non-Parametric GraphNet-Regularized
+        Representation of dMRI in Space and Time", Medical Image Analysis,
+        2017.
+    """
+    ux, uy, uz = us
+    x, y, z, _ = ind_mat.T
+    n_elem = ind_mat.shape[0]
+    LR = np.zeros((n_elem, n_elem))
+    for i in range(n_elem):
+        for k in range(i, n_elem):
+            if x[i] == x[k] and y[i] == y[k] and z[i] == z[k]:
+                LR[i, k] = LR[k, i] = (
+                    (1. / (ux * uy * uz)) * U_mat[x[i], x[k]] *
+                    U_mat[y[i], y[k]] * U_mat[z[i], z[k]]
+                )
+    return LR
+
+
+def part4_iso_reg_matrix_q(ind_mat, us):
+    """Partial spherical spatial Laplacian regularization matrix following the
+    equation below Eq. (C4) in [1].
+
+    References
+    ----------
+    .. [1] Fick, Rutger HJ, et al. "Non-Parametric GraphNet-Regularized
+        Representation of dMRI in Space and Time", Medical Image Analysis,
+        2017.
+    """
+    n_elem = ind_mat.shape[0]
+    LR = np.zeros((n_elem, n_elem))
+    for i in range(n_elem):
+        for k in range(i, n_elem):
+            if ind_mat[i, 0] == ind_mat[k, 0] and \
+               ind_mat[i, 1] == ind_mat[k, 1] and \
+               ind_mat[i, 2] == ind_mat[k, 2]:
+                ji = ind_mat[i, 0]
+           
```

Comment:
- unexpected parameter "gtab" in docstring. Missing parameter bvecs and qvals
- dtype is not used in this function. any reason ?